File: android/src/main/kotlin/com/example/easy_native/EasyNativeRouteRegistry.kt
package com.example.easy_native

import android.content.Context
import android.content.Intent

object EasyNativeRouteRegistry {
    const val EXTRA_ROUTE_NAME = "easy_native.route_name"
    const val EXTRA_PRESENTED = "easy_native.presented"

    private val nativeRoutes: MutableMap<String, (Context, Any?) -> Intent?> = mutableMapOf()

    fun registerNativeRoute(name: String, factory: (Context, Any?) -> Intent?) {
        val routeName = name.trim()
        if (routeName.isBlank()) {
            EasyNativeLogger.log(EasyNativeLogLevel.WARNING, "ignore empty native route registration")
            return
        }
        if (nativeRoutes.containsKey(routeName)) {
            EasyNativeLogger.log(EasyNativeLogLevel.WARNING, "override native route registration $routeName")
        } else {
            EasyNativeLogger.log(EasyNativeLogLevel.INFO, "register native route $routeName")
        }
        nativeRoutes[routeName] = factory
    }

    fun isNativeRoute(name: String): Boolean = nativeRoutes.containsKey(name)

    fun getNativeIntent(context: Context, name: String, args: Any?): Intent? {
        return nativeRoutes[name]?.invoke(context, args)?.apply {
            putExtra(EXTRA_ROUTE_NAME, name)
        }
    }
}

File: ios/Classes/EasyNativeRouteRegistry.swift
import UIKit

public final class EasyNativeRouteRegistry {
    public static let shared = EasyNativeRouteRegistry()

    private var nativeRoutes: [String: (Any?) -> UIViewController?] = [:]

    private init() {}

    public func registerNativeRoute(_ name: String, factory: @escaping (Any?) -> UIViewController?) {
        let routeName = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !routeName.isEmpty else {
            EasyNativeLogger.log(.warning, "ignore empty native route registration")
            return
        }
        if nativeRoutes[routeName] != nil {
            EasyNativeLogger.log(.warning, "override native route registration \(routeName)")
        } else {
            EasyNativeLogger.log(.info, "register native route \(routeName)")
        }
        nativeRoutes[routeName] = factory
    }

    public func isNativeRoute(_ name: String) -> Bool {
        return nativeRoutes[name] != nil
    }

    public func makeViewController(routeName: String, arguments: Any?) -> UIViewController? {
        guard let viewController = nativeRoutes[routeName]?(arguments) else {
            return nil
        }
        viewController.easyNativeRouteName = routeName
        viewController.easyNativeArguments = arguments
        return viewController
    }
}

private var easyNativeRouteNameKey: UInt8 = 0
private var easyNativeArgumentsKey: UInt8 = 0

extension UIViewController {
    public var easyNativeRouteName: String? {
        get {
            return objc_getAssociatedObject(self, &easyNativeRouteNameKey) as? String
        }
        set {
            objc_setAssociatedObject(
                self,
                &easyNativeRouteNameKey,
                newValue,
                .OBJC_ASSOCIATION_COPY_NONATOMIC
            )
        }
    }

    public var easyNativeArguments: Any? {
        get {
            return objc_getAssociatedObject(self, &easyNativeArgumentsKey)
        }
        set {
            objc_setAssociatedObject(
                self,
                &easyNativeArgumentsKey,
                newValue,
                .OBJC_ASSOCIATION_RETAIN_NONATOMIC
            )
        }
    }

    public func easyNativeTopMost() -> UIViewController {
        if let presented = presentedViewController {
            return presented.easyNativeTopMost()
        }
        if let navigationController = self as? UINavigationController,
           let visible = navigationController.visibleViewController {
            return visible.easyNativeTopMost()
        }
        if let tabBarController = self as? UITabBarController,
           let selected = tabBarController.selectedViewController {
            return selected.easyNativeTopMost()
        }
        return self
    }
}
